kernel/arch/riscv/paging.c

```diff
@@ -264,40 +951,268 @@ void *init_early_mapping(ulong fdt_paddr)
    return (void *)(kernel_pa - MB);
 }
 
+static inline bool
+in_big_page(pdir_t *pdir, void *vaddrp)
+{
+   page_table_t *pmd;
+   const ulong vaddr = (ulong) vaddrp;
+   page_t *e;
+
+   for (int level = RV_PAGE_LEVEL; level > 1; level--) {
+
+      e = &pdir->entries[PTE_INDEX(level, vaddr)];
+
+      if (!e->present)
+         return false;
+
+      /* 1 GB super big page, not we want*/
+      if (e->raw & _PAGE_LEAF)
+         return false;
+
+      pmd = PA_TO_LIN_VA(e->pfn << PAGE_SHIFT);
+      pdir = pmd;
+   }
+
+   e = &pdir->entries[PTE_INDEX(1, vaddr)];
+
+   return (e->present) && (e->raw & _PAGE_LEAF);
+}
+
+static void
+set_big_page_attr(pdir_t *pdir,
+                  void *vaddrp,
+                  ulong attr)
+{
+   page_table_t *pmd;
+   const ulong vaddr = (ulong) vaddrp;
+   ulong *pte;
+   page_t *e;
+
+   for (int level = RV_PAGE_LEVEL; level > 1; level--) {
+
+      e = &pdir->entries[PTE_INDEX(level, vaddr)];
+
+      ASSERT((e->present) && !(e->raw & _PAGE_LEAF));
+
+      pmd = PA_TO_LIN_VA(e->pfn << PAGE_SHIFT);
+      pdir = pmd;
+   }
+
+   pte = &pdir->entries[PTE_INDEX(1, vaddr)].raw;
+   *pte = (*pte & ~_PAGE_MTMASK ) | attr;
+   invalidate_page_hw(vaddr);
+}
+
+static void
+set_4kb_page_attr(pdir_t *pdir,
+                  void *vaddrp,
+                  ulong attr)
+{
+   page_table_t *pt;
+   ulong *pte;
+   const ulong vaddr = (ulong) vaddrp;
+
+   ASSERT(!(vaddr & OFFSET_IN_PAGE_MASK)); // the vaddr must be page-aligned
```

Comment:
Yeah, I used OFFSet_IN_PAGE_MASK because (PAGE_SIZE - 1) was too ugly. But a nice set of macros for check for alignment at L0, L1, L2 etc. would nice and could be used of any architecture (with a different definition of course).

Reply:
fixed!